src/collection/shared/generic_collector.py

```diff
@@ -140,17 +141,16 @@ def _rate_limit_calls(self) -> None:
         It ensures that the number of API calls does not exceed a predefined limit per second.
         If the limit is reached, it will pause the execution momentarily before allowing further API calls.
         """
-        # Check if the number of calls made in the last second exceeds the maximum limit.
-        while self.rate_limit <= len(self._call_timestamps):
-            # Calculate the time passed since the first recorded call.
-            time_passed = time.time() - self._call_timestamps[0]
+        now = time.time()
 
-            if time_passed > 1:
-                self._call_timestamps.pop(0)
-            else:
-                time.sleep(1 - time_passed)
+        # If we've reached the rate limit and the current time is within 1 second of the oldest call
+        if len(self._call_timestamps) == self.rate_limit and now - self._call_timestamps[0] < 1:
+            # Calculate the time to wait until making the next call
+            time_to_wait = 1 - (now - self._call_timestamps[0])
```

Comment:
Can you move this line before the `if` condition and reuse it then?

Reply:
done